File: agents/markdown-formatter-agent.ts
/**
 * Markdown Formatter Agent
 * 
 * Agent Skills:
 * - Detecta bullets e numeração misturados em parágrafos
 * - Adiciona quebras de linha antes de bullets/numeração
 * - Padroniza formatação MD antes de conversão para JSON
 * - Mantém histórico de ajustes realizados
 * - Valida formatação de saída
 */

export interface FormattingAdjustment {
  fieldName: string;
  originalText: string;
  adjustedText: string;
  changesApplied: string[];
  timestamp: Date;
}

export interface AgentMemory {
  adjustmentsHistory: FormattingAdjustment[];
  lastExecuted: Date;
  totalAdjustmentsMade: number;
}

export class MarkdownFormatterAgent {
  private memory: AgentMemory = {
    adjustmentsHistory: [],
    lastExecuted: new Date(),
    totalAdjustmentsMade: 0,
  };

  /**
   * Padrões a detectar em texto
   */
  private patterns = {
    // Bullet seguido de espaço e texto no mesmo parágrafo
    bulletMixed: /([^\n])\s+•\s+/g,
    // Numeração seguida de espaço e texto no mesmo parágrafo
    numberMixed: /([^\n])\s+(\d+\.)\s+/g,
    // Hífen como bullet seguido de espaço e texto no mesmo parágrafo
    dashMixed: /([^\n])\s+-\s+/g,
  };

  /**
   * Processa texto para adicionar quebras de linha antes de bullets/numeração
   */
  formatMarkdownText(text: string): string {
    if (!text) return text;

    let formattedText = text;
    const changes: string[] = [];

    // Detecta e adiciona quebra de linha antes de bullets
    if (this.patterns.bulletMixed.test(formattedText)) {
      formattedText = formattedText.replace(
        this.patterns.bulletMixed,
        '$1\n• '
      );
      changes.push('Adicionada quebra de linha antes de bullets (•)');
    }

    // Detecta e adiciona quebra de linha antes de numeração
    if (this.patterns.numberMixed.test(formattedText)) {
      formattedText = formattedText.replace(
        this.patterns.numberMixed,
        '$1\n$2 '
      );
      changes.push('Adicionada quebra de linha antes de numeração');
    }

    // Detecta e adiciona quebra de linha antes de hífens (quando usado como bullet)
    if (this.patterns.dashMixed.test(formattedText)) {
      // Mais conservador para não quebrar sentenças normais
      const dashLines = formattedText.split('\n');
      const processedLines = dashLines.map(line => {
        // Só aplica se o hífen aparecer no meio de um parágrafo já com bullets/numeração
        if (line.includes('•') || line.match(/\d+\./)) {
          return line.replace(/ - /g, '\n- ');
        }
        return line;
      });
      formattedText = processedLines.join('\n');
      if (changes[changes.length - 1] !== 'Adicionada quebra de linha antes de hífens') {
        changes.push('Adicionada quebra de linha antes de hífens (-)');
      }
    }

    return formattedText;
  }

  /**
   * Processa objeto JSON recursivamente, formatando todos os campos de texto
   */
  processJsonObject(obj: any, fieldName: string = 'root'): any {
    if (obj === null || obj === undefined) return obj;

    if (typeof obj === 'string') {
      const formatted = this.formatMarkdownText(obj);
      if (formatted !== obj) {
        this.recordAdjustment(fieldName, obj, formatted);
      }
      return formatted;
    }

    if (Array.isArray(obj)) {
      return obj.map((item, index) => 
        this.processJsonObject(item, `${fieldName}[${index}]`)
      );
    }

    if (typeof obj === 'object') {
      const processed: any = {};
      for (const key in obj) {
        if (obj.hasOwnProperty(key)) {
          processed[key] = this.processJsonObject(
            obj[key],
            `${fieldName}.${key}`
          );
        }
      }
      return processed;
    }

    return obj;
  }

  /**
   * Processa arquivo JSON completo
   */
  processSolutionData(data: any): any {
    const startTime = new Date();
    const processed = this.processJsonObject(data);
    this.memory.lastExecuted = startTime;
    
    return processed;
  }

  /**
   * Registra um ajuste no histórico
   */
  private recordAdjustment(
    fieldName: string,
    originalText: string,
    adjustedText: string
  ): void {
    const changes: string[] = [];

    if (originalText.match(this.patterns.bulletMixed)) {
      changes.push('Quebra de linha adicionada antes de bullets');
    }
    if (originalText.match(this.patterns.numberMixed)) {
      changes.push('Quebra de linha adicionada antes de numeração');
    }
    if (originalText.match(this.patterns.dashMixed)) {
      changes.push('Quebra de linha adicionada antes de hífens');
    }

    this.memory.adjustmentsHistory.push({
      fieldName,
      originalText: originalText.substring(0, 100),
      adjustedText: adjustedText.substring(0, 100),
      changesApplied: changes,
      timestamp: new Date(),
    });

    this.memory.totalAdjustmentsMade++;
  }

  /**
   * Retorna relatório de ajustes realizados
   */
  getReport(): AgentMemory {
    return {
      ...this.memory,
      lastExecuted: new Date(this.memory.lastExecuted),
    };
  }

  /**
   * Exibe resumo de ajustes no console
   */
  printReport(): void {
    console.log('\n=== MARKDOWN FORMATTER AGENT REPORT ===');
    console.log(`Total de ajustes realizados: ${this.memory.totalAdjustmentsMade}`);
    console.log(`Última execução: ${this.memory.lastExecuted}`);
    console.log('\nAjustes por campo:');
    
    this.memory.adjustmentsHistory.forEach((adjustment, index) => {
      console.log(`\n${index + 1}. Campo: ${adjustment.fieldName}`);
      console.log(`   Alterações: ${adjustment.changesApplied.join(', ')}`);
      console.log(`   Original: "${adjustment.originalText}..."`);
      console.log(`   Ajustado: "${adjustment.adjustedText}..."`);
    });
    
    console.log('\n=====================================\n');
  }

  /**
   * Exporta relatório em formato JSON
   */
  exportReport(): string {
    return JSON.stringify(this.getReport(), null, 2);
  }

  /**
   * Reseta o histórico de memória
   */
  resetMemory(): void {
    this.memory = {
      adjustmentsHistory: [],
      lastExecuted: new Date(),
      totalAdjustmentsMade: 0,
    };
  }
}

// Exporta singleton instance
export const markdownFormatterAgent = new MarkdownFormatterAgent();

File: agents/markdown_formatter_agent.py
"""
Markdown Formatter Agent para Python
Processa textos adicionando quebras de linha antes de bullets/numeração
"""
import re
from typing import Any, Dict, List
from datetime import datetime
from pathlib import Path


class MarkdownFormatterAgent:
    """Agent responsável por normalizar formatação de text com bullets/numeração"""
    
    def __init__(self):
        """Inicializa o agent com configurações padrão"""
        self.memory = {
            'adjustmentsHistory': [],
            'lastExecuted': None,
            'totalAdjustmentsMade': 0,
        }
        
        # Padrões regex para detectar problemas
        self.patterns = {
            # Bullet seguido de espaço e texto no mesmo parágrafo
            'bulletMixed': re.compile(r'([^\n])\s+•\s+'),
            # Numeração seguida de espaço e texto no mesmo parágrafo
            'numberMixed': re.compile(r'([^\n])\s+(\d+\.)\s+'),
            # Hífen como bullet seguido de espaço e texto no mesmo parágrafo
            'dashMixed': re.compile(r'([^\n])\s+-\s+(?!\s)'),
        }
        
        # Campos de texto que devem ser processados
        self.text_fields = [
            'descricao', 'objetivo', 'descricaoDetalhada', 'entrega',
            'beneficiosResultadosEsperados', 'estruturaMateriais',
            'responsabilidadeEmpresaDemandante', 'responsabilidadePrestadora',
            'perfilDesejadoPrestadora', 'observacoes', 'notas', 'conteudo',
            'texto', 'observacoesGerais', 'observacoesEspecificas'
        ]

    def format_markdown_text(self, text: str) -> str:
        """
        Processa texto para adicionar quebras de linha antes de bullets/numeração
        
        Args:
            text: Texto para processar
            
        Returns:
            Texto formatado com quebras de linha
        """
        if not text or not isinstance(text, str):
            return text

        formatted_text = text
        changes = []

        # Detecta e adiciona quebra de linha antes de bullets
        if self.patterns['bulletMixed'].search(formatted_text):
            formatted_text = self.patterns['bulletMixed'].sub(r'\1\n• ', formatted_text)
            changes.append('Quebra de linha adicionada antes de bullets (•)')

        # Detecta e adiciona quebra de linha antes de numeração
        if self.patterns['numberMixed'].search(formatted_text):
            formatted_text = self.patterns['numberMixed'].sub(r'\1\n\2 ', formatted_text)
            changes.append('Quebra de linha adicionada antes de numeração')

        # Detecta e adiciona quebra de linha antes de hífens
        if self.patterns['dashMixed'].search(formatted_text):
            formatted_text = self.patterns['dashMixed'].sub(r'\1\n- ', formatted_text)
            changes.append('Quebra de linha adicionada antes de hífens (-)')

        return formatted_text

    def process_json_object(self, obj: Any, field_name: str = 'root') -> Any:
        """
        Processa objeto recursivamente, formatando todos os campos de texto
        
        Args:
            obj: Objeto a processar (dict, list, string, etc)
            field_name: Nome do campo (para rastreamento)
            
        Returns:
            Objeto processado
        """
        if obj is None:
            return obj

        # Se é string, processar com o formatter
        if isinstance(obj, str):
            formatted = self.format_markdown_text(obj)
            if formatted != obj:
                self._record_adjustment(field_name, obj, formatted)
            return formatted

        # Se é lista, processar recursivamente
        if isinstance(obj, list):
            return [
                self.process_json_object(item, f'{field_name}[{idx}]')
                for idx, item in enumerate(obj)
            ]

        # Se é dict, processar recursivamente
        if isinstance(obj, dict):
            processed = {}
            for key, value in obj.items():
                processed[key] = self.process_json_object(
                    value,
                    f'{field_name}.{key}'
                )
            return processed

        return obj

    def process_solution_data(self, data: Any) -> Any:
        """
        Processa dados completos de uma solução
        
        Args:
            data: Dados da solução (dict)
            
        Returns:
            Dados processados
        """
        start_time = datetime.now()
        processed = self.process_json_object(data)
        self.memory['lastExecuted'] = start_time.strftime("%d/%m/%Y %H:%M:%S")
        
        return processed

    def _record_adjustment(self, field_name: str, original: str, adjusted: str) -> None:
        """Registra um ajuste no histórico"""
        changes = []

        if self.patterns['bulletMixed'].search(original):
            changes.append('Quebra de linha adicionada antes de bullets')
        if self.patterns['numberMixed'].search(original):
            changes.append('Quebra de linha adicionada antes de numeração')
        if self.patterns['dashMixed'].search(original):
            changes.append('Quebra de linha adicionada antes de hífens')

        adjustment = {
            'fieldName': field_name,
            'originalText': original[:100],
            'adjustedText': adjusted[:100],
            'changesApplied': changes,
            'timestamp': datetime.now().strftime("%d/%m/%Y %H:%M:%S"),
        }

        self.memory['adjustmentsHistory'].append(adjustment)
        self.memory['totalAdjustmentsMade'] += 1

    def get_report(self) -> Dict[str, Any]:
        """Retorna relatório de ajustes realizados"""
        return self.memory.copy()

    def print_report(self) -> None:
        """Exibe resumo de ajustes no console"""
        report = self.get_report()
        
        print('\n' + '=' * 70)
        print('📋 MARKDOWN FORMATTER AGENT REPORT')
        print('=' * 70)
        print(f"✅ Total de ajustes realizados: {report['totalAdjustmentsMade']}")
        print(f"⏱️  Última execução: {report['lastExecuted']}")
        
        if report['adjustmentsHistory']:
            print('\n📝 Ajustes por campo:')
            print('-' * 70)
            
            for idx, adjustment in enumerate(report['adjustmentsHistory'], 1):
                print(f"\n{idx}. Campo: {adjustment['fieldName']}")
                print(f"   Alterações: {', '.join(adjustment['changesApplied'])}")
                print(f"   Original:  \"{adjustment['originalText']}...\"")
                print(f"   Ajustado:  \"{adjustment['adjustedText']}...\"")
        else:
            print('\n✨ Nenhum ajuste necessário')
        
        print('\n' + '=' * 70 + '\n')

    def export_report(self) -> str:
        """Exporta relatório em formato JSON"""
        import json
        return json.dumps(self.get_report(), indent=2, ensure_ascii=False)

    def reset_memory(self) -> None:
        """Reseta o histórico de memória"""
        self.memory = {
            'adjustmentsHistory': [],
            'lastExecuted': None,
            'totalAdjustmentsMade': 0,
        }


# Instância singleton
_formatter_agent = None

def get_formatter_agent() -> MarkdownFormatterAgent:
    """Retorna instância singleton do agent"""
    global _formatter_agent
    if _formatter_agent is None:
        _formatter_agent = MarkdownFormatterAgent()
    return _formatter_agent
